### PR TITLE
Resolve method ambiguity based on deref level / if there are multiple impls of the same trait to the same type

### PR DESCRIPTION
**Pull request by vlad20012**

### CODE REVIEW

**Comment by matklad** [comment]
This could probably use a couple of `Testmark`s? :)

**Comment by vlad20012** [comment]
+ one more commit
+ use Teskmark

File: src/main/kotlin/org/rust/stdext/Collections.kt [hunk 1]
```diff
@@ -36,3 +36,13 @@ fun makeBitMask(bitToSet: Int): Int = 1 shl bitToSet
 
 fun <K, V1, V2> zipValues(map1: Map<K, V1>, map2: Map<K, V2>): List<Pair<V1, V2>> =
     map1.mapNotNull { (k, v1) -> map2[k]?.let { v2 -> Pair(v1, v2) } }
+
+inline fun <T> List<T>.singleOrFilter(predicate: (T) -> Boolean): List<T> = when {
+    size < 2 -> this
+    else -> filter(predicate)
+}
+
+inline fun <T> List<T>.singleOrLet(function: (List<T>) -> List<T>): List<T> = when {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/stdext/Collections.kt]: Can't think of a better name :(

File: src/test/kotlin/org/rust/lang/core/resolve/RsResolveTestBase.kt [hunk 2]
```diff
@@ -42,6 +43,9 @@ abstract class RsResolveTestBase : RsTestBase() {
         check(resolved == target)
     }
 
+    protected fun checkByCode(@Language("Rust") code: String, mark: Testmark) =
```
Comment (vlad20012) [inline src/test/kotlin/org/rust/lang/core/resolve/RsResolveTestBase.kt]: I think something like this make usage of `Testmark`s much more clear

**Comment by vlad20012** [comment]
+ more docs

**Comment by vlad20012** [comment]
bors r=matklad

**Comment by vlad20012** [comment]
There is a real problem with that test on windows. It fails locally somtimes...

**Pull request merged**